server/src/main/java/org/elasticsearch/action/DocWriteResponse.java

```diff
@@ -130,6 +130,22 @@ public DocWriteResponse(ShardId shardId, String id, long seqNo, long primaryTerm
         this.result = Objects.requireNonNull(result);
     }
 
+    // needed for deserialization
+    protected DocWriteResponse(ShardId shardId, StreamInput in) throws IOException {
+        super(in);
+        this.shardId = shardId;
+        if (in.getVersion().before(Version.V_8_0_0)) {
+            String type = in.readString();
+            assert MapperService.SINGLE_MAPPING_NAME.equals(type) : "Expected [_doc] but received [" + type + "]";
+        }
+        id = in.readString();
+        version = in.readZLong();
+        seqNo = in.readZLong();
+        primaryTerm = in.readVLong();
+        forcedRefresh = in.readBoolean();
+        result = Result.readFrom(in);
+    }
+
     // needed for deserialization
     protected DocWriteResponse(StreamInput in) throws IOException {
         super(in);
```

Comment:
Can we assert anything about `in.getVersion()` here? Not looked in detail, but I hope that this constructor will become obsolete after this change is complete.

Reply:
Not quite I think. We still have `TransportIndexAction` which still uses and logically needs the full deserialization on an `IndexResponse` as far as I can tell and won't go away in `8` right?

Reply:
Can't we get rid of that yet? It was deprecated over 3 years ago.

Reply:
Right in 8 we can ... opening a PR for that sec

Reply:
Urgh nevermind, technically we can remove this now but it's a far from trivial change to do so (we're still using that action all over the place in tests). I don't think I can do that in the short-term.

Reply:
Ok, can we mention `IndexAction` in this method's Javadoc so we keep track of the dependency.